src/Simulation/Simulators/OpenSystemsSimulator/Decompositions/IntrinsicInterface.qs

```diff
@@ -0,0 +1,210 @@
+// M.Q.Intrinsic, making it easier to forward intrinsic definitions from the
+// simulator.
+
+namespace Microsoft.Quantum.Experimental.Intrinsic {
+    open Microsoft.Quantum.Experimental.Native as Native;
+    open Microsoft.Quantum.Experimental.Decompositions;
+
+    operation H(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.H(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                H(target);
+            } else {
+                ApplyControlledH(controls, target);
+            }
+        }
+    }
+
+    operation X(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.X(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                X(target);
+            } elif Length(controls) == 1 {
+                Native.CNOT(controls[0], target);
+            } elif Length(controls) == 2 {
+                // Forward to decomposition of CCZ.
+                within {
+                    H(target);
+                } apply {
+                    Controlled Z(controls, target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled X, (controls, target));
+            }
+        }
+    }
+
+    operation Y(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.Y(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                Y(target);
+            } elif Length(controls) == 1 {
+                within {
+                    H(target);
+                    S(target);
+                } apply {
+                    Native.CNOT(controls[0], target);
+                }
+            } elif Length(controls) == 2 {
+                within {
+                    // TODO: double-check this one...!
+                    within {
+                        H(target);
+                    } apply {
+                        S(target);
+                    }
+                } apply {
+                    Controlled Z(controls, target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled Y, (controls, target));
+            }
+        }
+    }
+
+    operation Z(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.Z(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                Z(target);
+            } elif Length(controls) == 1 {
+                within {
+                    H(target);
+                } apply {
+                    Native.CNOT(controls[0], target);
+                }
+            } elif Length(controls) == 1 {
+                // [Page 15 of arXiv:1206.0758v3](https://arxiv.org/pdf/1206.0758v3.pdf#page=15)
+                Adjoint T(controls[0]);
+                Adjoint T(controls[1]);
+                Controlled X([target], controls[0]);
```

Comment:
It's a very minor optimization, but consider using `Native.CNOT` here instead of `Controlled X` (and in other decompositions) as that can avoid some extra overhead in invoking the specailization code and should be safe to use here since `Native.CNOT` is already explicitly implemented as self-adjoint.

Reply:
Good point, went on and applied that optimization.